### PR TITLE
Olive Command Line Tools

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
This PR adds command line tools for Olive that can be called using `olive` or `python -m olive` command.
- Previous commands `python -m olive.workflows.run` and `python -m olive.platform_sdk.qualcomm.configure` are deprecated and will emit a FutureWarning.
- All cli commands are implemented in the `olive.cli` sub-module. These must handle the command parsing and, in some cases, the implementation of the command too.
- Removed `olive.workflows.snpe` sub-module since the code is outdated and obsolete.
- Some doc fixes where we need to use "``" instead of "`" for code formatting.
- Documentation is automatic using the [`sphinx-argparse` extension](https://sphinx-argparse.readthedocs.io/en/stable/index.html):
image

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
    - `olive` command line tool has been added. 
    - Use `olive run` or `python -m olive` instead of `python -m olive.workflows.run` which has been deprecated.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/cli/configure_qualcomm_sdk.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            type=str,
+            help="Qualcomm SDK: snpe or qnn",
+            required=True,
+            choices=["snpe", "qnn"],
+        )
+
+        sub_parser.set_defaults(func=ConfigureQualcommSDKCommand)
+
+    def run(self):
+        from olive.platform_sdk.qualcomm.configure import configure
```
[the remark this answers is not in the document]
Comment (jambayk) [reply olive/cli/configure_qualcomm_sdk.py]: Fixed

File: olive/cli/launcher.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import sys
+from argparse import ArgumentParser
+from warnings import warn
+
+from olive.cli.configure_qualcomm_sdk import ConfigureQualcommSDKCommand
```
[the remark this answers is not in the document]
Comment (jambayk) [reply olive/cli/launcher.py:9]: Fixed

File: olive/cli/run.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        sub_parser.set_defaults(func=WorkflowRunCommand)
+
+    def run(self):
+        from olive.common.utils import set_tempdir
+        from olive.workflows import run as olive_run
+
+        set_tempdir(self.args.tempdir)
+
+        var_args = vars(self.args)
+        del var_args["func"], var_args["tempdir"]
```
Comment (shaahji) [inline olive/cli/run.py:41]: Where is `func` coming from? It is not part of the args list.

**Review by shaahji — CHANGES_REQUESTED** [review]
While you are touching so many of the documentation files, I would recommend adding the language for the commands as well. A lot of them are missing.

Instead of wrapping \``` .... \``` it should be \```*sh* ..... \```. Or optionally yse `bash`

File: docs/source/tutorials/configure_auto_optimizer.rst [hunk 4]
```diff
@@ -38,7 +38,7 @@ Config Fields:
         - If set to True, Auto Optimizer will be disabled and user need to provide passes combination manually.
 
     3. precision[optional[str]]: default None.
-        - The precision of output model. If user does not set the precision of output model, it will be determined by above optimization factors. We supports "fp32", "fp16" and "int8" output precision for now.
+        - The precision of output model. If user does not set the precision of output model, it will be determined by above optimization factors. We support "fp32", "fp16" and "int8" output precision for now.
```
[the remark this answers is not in the document]
Comment (jambayk) [reply docs/source/tutorials/configure_auto_optimizer.rst]: Done!

File: docs/source/tutorials/azureml_scripts.md [hunk 5]
```diff
@@ -42,19 +42,21 @@ This Python script provides a command-line interface for managing compute resour
 
 #### Usage
 
-You can run the script to create an AzureML compute instance from the command line like this:
+You can use ``olive manage-aml-compute`` command line tool create an AzureML compute instance from the command line like this:
```
[the remark this answers is not in the document]
Comment (jambayk) [reply docs/source/tutorials/azureml_scripts.md]: Done!

File: olive/cli/run.py (hunk 3 above)
Comment (jambayk) [reply olive/cli/run.py:41]: It gets added here https://github.com/microsoft/Olive/blob/cf03e92a7cfb215447cfff6507f1a99774fd9122/olive/cli/run.py#L32

**Review by shaahji — APPROVED** [review]

**Pull request merged**